common/services/user.js

```diff
@@ -107,8 +108,32 @@ function getNomisLocations(token) {
     )
 }
 
+async function populateSupplierLocations(user) {
+  const { supplierId, permissions } = user
+
+  if (supplierId) {
+    user.locations = await referenceDataService.getLocationsBySupplierId(
```

Comment:
I wonder if it would be better to return the locations here rather than mutate the `user` object like it was previously doing?

Reply:
Yeah, I wondered about that too. I decided to go with this method because I wasn't sure whether the return value should be `null` or `user.locations` if the locations isn't being changed (i.e. the user isn't a supplier) and it felt a bit messy. I'll have another think about what we could do here, because I definitely agree a function which mutates the object isn't great either.

Reply:
Yes, I see what you mean. I was also wondering if it could have been incorporated into the [`User`](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/a732c065c08d0b8a288f7a0fba63d16c8b3ee208/common/lib/user.js#L13) class similar to [how permissions are set](https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/blob/a732c065c08d0b8a288f7a0fba63d16c8b3ee208/common/lib/user.js#L23) but it would involve making that async which could make things a little more complex...?

I wonder if the promises in the auth middleware could actually be updated to contain the logic as that's almost where it should be determined. In these cases where the user is a supplier or has a particular role we wouldn't need to carry out the other locations call either, right?

Reply:
I agree this is definitely the way forward. I'm concerned that making this change in this PR would make it quite a big change, and I think I'd prefer to have it at least as a separate PR, but maybe even a separate ticket.

I've had a little play around how we could do it differently. The main problem is that because `user.permissions` is done within the `User` model, we need the `user` instance available before we can try to fetch the supplier locations. But for the `user` to exist, we need to have already tried to fetch the `locations` so we can construct a `User` instance. I definitely agree this is problem and ripe for refactoring, but I'm worried doing it in this PR is too much.

At least with the PR as it is the bug itself should be fixed and the behaviour of _how_ the supplier locations hasn't changed, it's just the _when_. Does that make sense?

Reply:
Yep, I agree with all of this. I think that's a really good approach, and as you say, the main goal was to prevent the locations being lost when a user's session is regenerated without them visiting the locations controller :)

Maybe we can add a GH issue with some of these suggestions and can pick it up when/if there is some capacity?

Reply:
Sounds good, I'll write those issues this afternoon!

Reply:
I've written up https://github.com/ministryofjustice/hmpps-book-secure-move-frontend/issues/1805.